### PR TITLE
[ENH] Improved data preparation for LLM finetuning

### PR DESCRIPTION
**Pull request by paxcema**
## Description

This PR introduces some common data preparation utilities for fine-tuning LLMs that operate following the ChatCompletion format from OpenAI. Currently, these coupled versions of these methods are actively being used in our integrations for Anyscale and OpenAI (hence indirectly by LangChain, too).

Added tests for all of these, so that we can use them on any other LLM engines.

## Type of change

- [x] ⚡ New feature (non-breaking change which adds functionality). Refactor.

## Verification Process

To ensure the changes are working as expected:

 - [x]   Test Location: added unit tests, which pass.

## Additional Media:

- [ ] I have attached a brief loom video or screenshots showcasing the new functionality or change.

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have appropriately commented on my code, especially in complex areas.
- [ ] Necessary documentation updates are either made or tracked in issues.
- [x] Relevant unit and integration tests are updated or added.

### CODE REVIEW

File: tests/unit/test_llm_utils.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                {"role": "user", "content": "hi"},
+                {"role": "assistant", "content": None},  # should always be a string
+            ],
+
+            # invalid - no assistant in the chat
+            [
+                {"role": "user", "content": "hi"},
+            ],
+        ]
+
```
Comment (dusvyat) [inline tests/unit/test_llm_utils.py:112]: minor nit: perhaps we can make invalid_chats and valid_chats a fixture. Stylistic, doesn't block merge

**Review by dusvyat — APPROVED** [review]
LGTM, see minor point

File: mindsdb/integrations/handlers/anyscale_endpoints_handler/anyscale_endpoints_handler.py [hunk 2]
```diff
@@ -119,63 +120,35 @@ def _check_ft_cols(df, cols):
...
-            chats.append({'messages': chat})
+    # TODO: add unit tests
+    @staticmethod
+    def _prepare_ft_jsonl(df, temp_storage_path, temp_filename, _, test_size=0.2):
+        # 1. format data
+        chats = ft_chat_formatter(df)
 
+        # 2. split chats in training and validation subsets
         series = pd.Series(chats)
         if len(series) < 20 * 2:
```
Comment (pedrofluxa) [inline mindsdb/integrations/handlers/anyscale_endpoints_handler/anyscale_endpoints_handler.py:131]: This seems a little arbitrary; shouldn't the length of the dataset depend on how large the model is? I would at least leave that "20" as a keyword parameter with a default value.

**Review by pedrofluxa — APPROVED** [review]
Left a comment on a hard-coded value. I am approving this PR as it should work as it is, but I strongly suggest to refactor the code a little bit.

File: mindsdb/integrations/handlers/anyscale_endpoints_handler/anyscale_endpoints_handler.py (hunk 2 above)
Comment (paxcema) [reply mindsdb/integrations/handlers/anyscale_endpoints_handler/anyscale_endpoints_handler.py:131]: Good shout, I'll change the row number to be customizable. As for why: Anyscale endpoints has a non-documented cutoff at 20 -- if the validation dataset is smaller than that, the entire process fails. I'll add this info. as a comment.

Could you expand on how I should go about this one?
> I strongly suggest to refactor the code a little bit.

**Comment by chandrevdw31** [comment]
Unable to Finetune as it says False 

image

**Comment by paxcema** [comment]
@chandrevdw31 status is marked as `generating` in the screenshot above, and these finetune runs can easily take hours so this is normal. The model cannot be `active` before its status is `complete`. 

If the user reports an actual error in the model record, please ask them to open a bug with it.

**Pull request merged**